tests/blockchain.test.js

```diff
@@ -1,99 +1,136 @@
-    		const validTx = createSignedTx();
-    		validTx.toAddress = null;
+    it('should fail when not having enough balance', function() {
+      const tx1 = createSignedTx();
+      assert.throws(() => { blockchain.addTransaction(tx1); }, Error);
+    });
+  });
 
-    		assert.throws(() => { blockchain.addTransaction(validTx) }, Error);
-    	});
+  describe('wallet balance', function() {
+    it('should give mining rewards', function() {
+      let blockchain = createBCWithMined();
+      const validTx = createSignedTx();
+      blockchain.addTransaction(validTx);
+      blockchain.addTransaction(validTx);
 
-    	it('should fail when tx is not valid', function(){
-    		const validTx = createSignedTx();
-    		validTx.amount = 1000;
+      blockchain.minePendingTransactions('b2');
 
-    		assert.throws(() => { blockchain.addTransaction(validTx) }, Error);
-    	});
+      assert.strict.equal(blockchain.getBalanceOfAddress('b2'), 100);
     });
 
-    describe('wallet balance', function(){
-    	it('should give mining rewards', function(){
-    		const validTx = createSignedTx();
-    		blockchain.addTransaction(validTx);
-    		blockchain.addTransaction(validTx);
+    it('should correctly reduce wallet balance', function() {
+      const walletAddr = signingKey.getPublic('hex');
+      blockchain = createBlockchainWithTx();
 
-    		blockchain.minePendingTransactions("b2");
+      blockchain.minePendingTransactions(walletAddr);
+      assert.strict.equal(blockchain.getBalanceOfAddress(walletAddr), 180);
+    });
+  });
 
-    		assert.equal(blockchain.getBalanceOfAddress("b2"), 100);
-    	});
+  describe('helper functions', function() {
+    it('should correctly set first block to genesis block', function() {
+      assert.strict.deepEqual(blockchain.chain[0], blockchain.createGenesisBlock());
+    });
+  });
 
-    	it('should correctly reduce wallet balance', function(){
-    		const walletAddr = signingKey.getPublic('hex');
-    		blockchain = createBlockchainWithTx();
+  describe('isChainValid', function() {
+    it('should return true if no tampering', function() {
+      blockchain = createBlockchainWithTx();
+      assert(blockchain.isChainValid());
+    });
 
-    		blockchain.minePendingTransactions(walletAddr);
-    		assert.equal(blockchain.getBalanceOfAddress(walletAddr), 80);
-    	});
+    it('should fail when genesis block has been tampered with', function() {
+      blockchain.chain[0].timestamp = 39708;
+      assert(!blockchain.isChainValid());
     });
 
-    describe('helper functions', function(){
-    	it('should correctly set first block to genesis block', function(){
-    		assert.deepEqual(blockchain.chain[0], blockchain.createGenesisBlock());
-    	});
+    it('should fail when a tx is invalid', function() {
+      blockchain = createBlockchainWithTx();
+      blockchain.chain[1].transactions[0].amount = 897397;
+      assert(!blockchain.isChainValid());
```

Comment:
somehow this is not really working how it should, anyone know why?
(returns because .calculateHash, should return because .hasValidTransactions)

Reply:
and btw: is ".hasValidTransactions" needed? wouldnt it be catched with ".calculateHash" too?

Reply:
Lets keep `hasValidTransactions` in for now. I want to be able to merge this PR.

I don't really understand your first comment. Can you elaborate a bit? 

All tests seem to be passing fine. And yes, the `isValid` method will always return true if it's a transaction that was generated by the system.

Reply:
> I don't really understand your first comment. Can you elaborate a bit?

its old and i dont really know the code anymore (i didnt use it much...) so i dont really know what i meant there

but from what i recall is: the test passes, but it shouldnt have

Reply:
update: i think i fully know what i meant again:
I meant that it "failed" (what is expected in this test), what is right, but it failed because of the check for "calculateHash", thats why i later removed the check for "hasValidTransactions", sorry for the confusion